# Consumer Complaints Resolution


Consumer complaint resolution is important to any business. In this particular case we have been given detailed
consumer complaints along with whether consumer disputed with the conclusion. If we are able to predict this,
consumer who is more likely to dispute a conclusion can be given more attention as to how the complaints are
handled as well as how persuasively the final conlusions are conveyed to them.
Your target here is to build prediction model for column "Consumer disputed"

In [71]:
import pandas as pd
import numpy as np

In [72]:
train_data=pd.read_csv(r'C:\Users\chandrakanth.seela\Downloads\P1 Data\Consumer_Complaints_train.csv')

In [73]:
test_data=pd.read_csv(r'C:\Users\chandrakanth.seela\Downloads\P1 Data\Consumer_Complaints_test_share.csv')

In [74]:
train_data.head()

Date received                  Product                  Sub-product  \
0    2014-05-15              Credit card                          NaN   
1    2014-09-18  Bank account or service  (CD) Certificate of deposit   
2    2014-03-13         Credit reporting                          NaN   
3    2015-07-17              Credit card                          NaN   
4    2014-11-20              Credit card                          NaN   

                                      Issue       Sub-issue  \
0                         Billing statement             NaN   
1  Making/receiving payments, sending money             NaN   
2    Incorrect information on credit report  Account status   
3                         Billing statement             NaN   
4                         Transaction issue             NaN   

                        Consumer complaint narrative  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  My credit card statement from US Bank, XXXX. X...   
4                                                NaN   

                            Company public response                Company  \
0                                               NaN  Wells Fargo & Company   
1                                               NaN      Santander Bank US   
2                                               NaN                Equifax   
3  Company chooses not to provide a public response           U.S. Bancorp   
4                                               NaN        Bank of America   

  State ZIP code            Tags Consumer consent provided? Submitted via  \
0    MI    48342  Older American                        NaN           Web   
1    PA    18042             NaN                        NaN      Referral   
2    CA    92427             NaN                        NaN      Referral   
3    GA    305XX  Older American           Consent provided           Web   
4    MA    02127             NaN                        NaN           Web   

  Date sent to company     Company response to consumer Timely response?  \
0           2014-05-16          Closed with explanation              Yes   
1           2014-09-24                           Closed              Yes   
2           2014-04-03  Closed with non-monetary relief              Yes   
3           2015-07-17      Closed with monetary relief              Yes   
4           2014-11-28          Closed with explanation              Yes   

  Consumer disputed?  Complaint ID  
0                 No        856103  
1                 No       1034666  
2                 No        756363  
3                 No       1474177  
4                 No       1132572

In [75]:
test_data.head()

Date received                  Product  \
0    2014-01-18  Bank account or service   
1    2016-03-31          Debt collection   
2    2012-03-08                 Mortgage   
3    2016-01-07         Credit reporting   
4    2013-08-23                 Mortgage   

                              Sub-product  \
0      Cashing a check without an account   
1                             Credit card   
2  Conventional adjustable mortgage (ARM)   
3                                     NaN   
4                            FHA mortgage   

                                      Issue  \
0                  Deposits and withdrawals   
1     Cont'd attempts collect debt not owed   
2  Loan servicing, payments, escrow account   
3  Unable to get credit report/credit score   
4  Loan modification,collection,foreclosure   

                                Sub-issue Consumer complaint narrative  \
0                                     NaN                          NaN   
1                           Debt was paid                          NaN   
2                                     NaN                          NaN   
3  Problem getting report or credit score                          NaN   
4                                     NaN                          NaN   

                            Company public response  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3  Company chooses not to provide a public response   
4                                               NaN   

                                  Company State ZIP code            Tags  \
0                         Bank of America    CA    95691             NaN   
1          National Credit Adjusters, LLC    FL    32086             NaN   
2                   Wells Fargo & Company    CA    94618             NaN   
3  TransUnion Intermediate Holdings, Inc.    FL    33584  Older American   
4                         Bank of America    FL    33543             NaN   

  Consumer consent provided? Submitted via Date sent to company  \
0                        NaN           Web           2014-01-17   
1       Consent not provided           Web           2016-03-31   
2                        NaN           Web           2012-03-09   
3                        NaN   Postal mail           2016-01-12   
4                        NaN           Web           2013-08-23   

      Company response to consumer Timely response?  Complaint ID  
0          Closed with explanation              Yes        675956  
1          Closed with explanation              Yes       1858795  
2            Closed without relief              Yes         32637  
3  Closed with non-monetary relief              Yes       1731374  
4          Closed with explanation              Yes        501487

In [76]:
print(f'training data has {train_data.shape[0]} rows and {train_data.shape[1]} columns')
print(f'test data has {test_data.shape[0]} rows and {test_data.shape[1]} columns')

training data has 478421 rows and 18 columns
test data has 119606 rows and 17 columns


Since the features for training data and test data must be same,lets combine the training data and test data so that the data pre-processing part can be done simultaneously for both the training data amd test data

In [77]:
#creating the response column for test set and filling it with null values()
test_data['Consumer disputed?']=np.nan

In [78]:
train_data['data']='train'
test_data['data']='test'

In [79]:
test_data.head()

Date received                  Product  \
0    2014-01-18  Bank account or service   
1    2016-03-31          Debt collection   
2    2012-03-08                 Mortgage   
3    2016-01-07         Credit reporting   
4    2013-08-23                 Mortgage   

                              Sub-product  \
0      Cashing a check without an account   
1                             Credit card   
2  Conventional adjustable mortgage (ARM)   
3                                     NaN   
4                            FHA mortgage   

                                      Issue  \
0                  Deposits and withdrawals   
1     Cont'd attempts collect debt not owed   
2  Loan servicing, payments, escrow account   
3  Unable to get credit report/credit score   
4  Loan modification,collection,foreclosure   

                                Sub-issue Consumer complaint narrative  \
0                                     NaN                          NaN   
1                           Debt was paid                          NaN   
2                                     NaN                          NaN   
3  Problem getting report or credit score                          NaN   
4                                     NaN                          NaN   

                            Company public response  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3  Company chooses not to provide a public response   
4                                               NaN   

                                  Company State ZIP code            Tags  \
0                         Bank of America    CA    95691             NaN   
1          National Credit Adjusters, LLC    FL    32086             NaN   
2                   Wells Fargo & Company    CA    94618             NaN   
3  TransUnion Intermediate Holdings, Inc.    FL    33584  Older American   
4                         Bank of America    FL    33543             NaN   

  Consumer consent provided? Submitted via Date sent to company  \
0                        NaN           Web           2014-01-17   
1       Consent not provided           Web           2016-03-31   
2                        NaN           Web           2012-03-09   
3                        NaN   Postal mail           2016-01-12   
4                        NaN           Web           2013-08-23   

      Company response to consumer Timely response?  Complaint ID  \
0          Closed with explanation              Yes        675956   
1          Closed with explanation              Yes       1858795   
2            Closed without relief              Yes         32637   
3  Closed with non-monetary relief              Yes       1731374   
4          Closed with explanation              Yes        501487   

   Consumer disputed?  data  
0                 NaN  test  
1                 NaN  test  
2                 NaN  test  
3                 NaN  test  
4                 NaN  test

In [80]:
combined_data=pd.concat([train_data,test_data],axis=0,sort=False)

In [81]:
combined_data.head()

Date received                  Product                  Sub-product  \
0    2014-05-15              Credit card                          NaN   
1    2014-09-18  Bank account or service  (CD) Certificate of deposit   
2    2014-03-13         Credit reporting                          NaN   
3    2015-07-17              Credit card                          NaN   
4    2014-11-20              Credit card                          NaN   

                                      Issue       Sub-issue  \
0                         Billing statement             NaN   
1  Making/receiving payments, sending money             NaN   
2    Incorrect information on credit report  Account status   
3                         Billing statement             NaN   
4                         Transaction issue             NaN   

                        Consumer complaint narrative  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  My credit card statement from US Bank, XXXX. X...   
4                                                NaN   

                            Company public response                Company  \
0                                               NaN  Wells Fargo & Company   
1                                               NaN      Santander Bank US   
2                                               NaN                Equifax   
3  Company chooses not to provide a public response           U.S. Bancorp   
4                                               NaN        Bank of America   

  State ZIP code            Tags Consumer consent provided? Submitted via  \
0    MI    48342  Older American                        NaN           Web   
1    PA    18042             NaN                        NaN      Referral   
2    CA    92427             NaN                        NaN      Referral   
3    GA    305XX  Older American           Consent provided           Web   
4    MA    02127             NaN                        NaN           Web   

  Date sent to company     Company response to consumer Timely response?  \
0           2014-05-16          Closed with explanation              Yes   
1           2014-09-24                           Closed              Yes   
2           2014-04-03  Closed with non-monetary relief              Yes   
3           2015-07-17      Closed with monetary relief              Yes   
4           2014-11-28          Closed with explanation              Yes   

  Consumer disputed?  Complaint ID   data  
0                 No        856103  train  
1                 No       1034666  train  
2                 No        756363  train  
3                 No       1474177  train  
4                 No       1132572  train

In [82]:
combined_data.tail()

Date received           Product                            Sub-product  \
119601    2012-09-15       Credit card                                    NaN   
119602    2016-07-21  Credit reporting                                    NaN   
119603    2015-06-09   Debt collection  Other (i.e. phone, health club, etc.)   
119604    2014-02-26   Debt collection                            Credit card   
119605    2015-02-11   Debt collection  Other (i.e. phone, health club, etc.)   

                                           Issue  \
119601                Closing/Cancelling account   
119602  Credit reporting company's investigation   
119603           Disclosure verification of debt   
119604           Disclosure verification of debt   
119605     Cont'd attempts collect debt not owed   

                                       Sub-issue  \
119601                                       NaN   
119602  No notice of investigation status/result   
119603      Not given enough info to verify debt   
119604      Not given enough info to verify debt   
119605                          Debt is not mine   

                             Consumer complaint narrative  \
119601                                                NaN   
119602  On XXXX/XXXX/16 a dispute and validation of a ...   
119603                                                NaN   
119604                                                NaN   
119605                                                NaN   

                                  Company public response  \
119601                                                NaN   
119602  Company has responded to the consumer and the ...   
119603   Company chooses not to provide a public response   
119604                                                NaN   
119605                                                NaN   

                                     Company State ZIP code           Tags  \
119601                  JPMorgan Chase & Co.    IL    60538            NaN   
119602                              Experian    CA    956XX            NaN   
119603  FAIR COLLECTIONS & OUTSOURCING, INC.    CA    90803  Servicemember   
119604                  Encore Capital Group    PA    17764            NaN   
119605    AllianceOne Recievables Management    CA    92103            NaN   

       Consumer consent provided? Submitted via Date sent to company  \
119601                        NaN           Web           2012-09-21   
119602           Consent provided           Web           2016-07-21   
119603                        NaN         Phone           2015-06-12   
119604                        NaN         Phone           2014-02-27   
119605                        NaN           Web           2015-02-11   

       Company response to consumer Timely response? Consumer disputed?  \
119601  Closed with monetary relief              Yes                NaN   
119602      Closed with explanation              Yes                NaN   
119603      Closed with explanation              Yes                NaN   
119604      Closed with explanation              Yes                NaN   
119605      Closed with explanation              Yes                NaN   

        Complaint ID  data  
119601        153482  test  
119602       2023523  test  
119603       1413678  test  
119604        732458  test  
119605       1235226  test

In [83]:
combined_data.dtypes

Date received                   object
Product                         object
Sub-product                     object
Issue                           object
Sub-issue                       object
Consumer complaint narrative    object
Company public response         object
Company                         object
State                           object
ZIP code                        object
Tags                            object
Consumer consent provided?      object
Submitted via                   object
Date sent to company            object
Company response to consumer    object
Timely response?                object
Consumer disputed?              object
Complaint ID                     int64
data                            object
dtype: object

It is obvious that there are no numeric features in the data set.Although Complaint ID is a numeric column it does not make sense to consider ID as a feature as we cannot infer about anything about the complaint based on its ID

There are two columns with date format in the data set which are present as object type.Let's convert them into proper date format

In [84]:
#converting the columns Date received and Date sent to company to proper date format
combined_data['Date received']=pd.to_datetime(combined_data['Date received'],infer_datetime_format=True)
combined_data['Date sent to company']=pd.to_datetime(combined_data['Date sent to company'],infer_datetime_format=True)

In [85]:
combined_data.dtypes

Date received                   datetime64[ns]
Product                                 object
Sub-product                             object
Issue                                   object
Sub-issue                               object
Consumer complaint narrative            object
Company public response                 object
Company                                 object
State                                   object
ZIP code                                object
Tags                                    object
Consumer consent provided?              object
Submitted via                           object
Date sent to company            datetime64[ns]
Company response to consumer            object
Timely response?                        object
Consumer disputed?                      object
Complaint ID                             int64
data                                    object
dtype: object

Let's create a new feature as the difference between Date received and Date sent to company

In [86]:
#creating a new column day_difference out of two columns Date sent to company and Date received
combined_data['day_difference']=pd.to_numeric(combined_data['Date sent to company'] - combined_data['Date received'])

In [87]:
combined_data.head()

Date received                  Product                  Sub-product  \
0    2014-05-15              Credit card                          NaN   
1    2014-09-18  Bank account or service  (CD) Certificate of deposit   
2    2014-03-13         Credit reporting                          NaN   
3    2015-07-17              Credit card                          NaN   
4    2014-11-20              Credit card                          NaN   

                                      Issue       Sub-issue  \
0                         Billing statement             NaN   
1  Making/receiving payments, sending money             NaN   
2    Incorrect information on credit report  Account status   
3                         Billing statement             NaN   
4                         Transaction issue             NaN   

                        Consumer complaint narrative  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  My credit card statement from US Bank, XXXX. X...   
4                                                NaN   

                            Company public response                Company  \
0                                               NaN  Wells Fargo & Company   
1                                               NaN      Santander Bank US   
2                                               NaN                Equifax   
3  Company chooses not to provide a public response           U.S. Bancorp   
4                                               NaN        Bank of America   

  State ZIP code            Tags Consumer consent provided? Submitted via  \
0    MI    48342  Older American                        NaN           Web   
1    PA    18042             NaN                        NaN      Referral   
2    CA    92427             NaN                        NaN      Referral   
3    GA    305XX  Older American           Consent provided           Web   
4    MA    02127             NaN                        NaN           Web   

  Date sent to company     Company response to consumer Timely response?  \
0           2014-05-16          Closed with explanation              Yes   
1           2014-09-24                           Closed              Yes   
2           2014-04-03  Closed with non-monetary relief              Yes   
3           2015-07-17      Closed with monetary relief              Yes   
4           2014-11-28          Closed with explanation              Yes   

  Consumer disputed?  Complaint ID   data    day_difference  
0                 No        856103  train    86400000000000  
1                 No       1034666  train   518400000000000  
2                 No        756363  train  1814400000000000  
3                 No       1474177  train                 0  
4                 No       1132572  train   691200000000000

In [88]:
# Dropping the colmnumns Date received and Date sent to company
combined_data.drop(['Date received','Date sent to company'],axis=1,inplace=True)

In [89]:
combined_data.head()

Product                  Sub-product  \
0              Credit card                          NaN   
1  Bank account or service  (CD) Certificate of deposit   
2         Credit reporting                          NaN   
3              Credit card                          NaN   
4              Credit card                          NaN   

                                      Issue       Sub-issue  \
0                         Billing statement             NaN   
1  Making/receiving payments, sending money             NaN   
2    Incorrect information on credit report  Account status   
3                         Billing statement             NaN   
4                         Transaction issue             NaN   

                        Consumer complaint narrative  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  My credit card statement from US Bank, XXXX. X...   
4                                                NaN   

                            Company public response                Company  \
0                                               NaN  Wells Fargo & Company   
1                                               NaN      Santander Bank US   
2                                               NaN                Equifax   
3  Company chooses not to provide a public response           U.S. Bancorp   
4                                               NaN        Bank of America   

  State ZIP code            Tags Consumer consent provided? Submitted via  \
0    MI    48342  Older American                        NaN           Web   
1    PA    18042             NaN                        NaN      Referral   
2    CA    92427             NaN                        NaN      Referral   
3    GA    305XX  Older American           Consent provided           Web   
4    MA    02127             NaN                        NaN           Web   

      Company response to consumer Timely response? Consumer disputed?  \
0          Closed with explanation              Yes                 No   
1                           Closed              Yes                 No   
2  Closed with non-monetary relief              Yes                 No   
3      Closed with monetary relief              Yes                 No   
4          Closed with explanation              Yes                 No   

   Complaint ID   data    day_difference  
0        856103  train    86400000000000  
1       1034666  train   518400000000000  
2        756363  train  1814400000000000  
3       1474177  train                 0  
4       1132572  train   691200000000000

Let's deal with the missing values in the training data...

In [90]:
missing_count=(combined_data.loc[combined_data['data']=='train']).isnull().sum().sort_values(ascending=False)
missing_percentage=(combined_data.loc[combined_data['data']=='train']).isnull().sum().sort_values(ascending=False)/len((combined_data.loc[combined_data['data']=='train']))
pd.concat([missing_count,missing_percentage],axis=1,keys=['missing_count','missing_percentage'])

missing_count  missing_percentage
Tags                                 411215            0.859525
Consumer complaint narrative         403327            0.843038
Company public response              388029            0.811062
Consumer consent provided?           342934            0.716804
Sub-issue                            292625            0.611647
Sub-product                          138473            0.289438
ZIP code                               3848            0.008043
State                                  3839            0.008024
Issue                                     0            0.000000
Company                                   0            0.000000
day_difference                            0            0.000000
data                                      0            0.000000
Submitted via                             0            0.000000
Company response to consumer              0            0.000000
Timely response?                          0            0.000000
Consumer disputed?                        0            0.000000
Complaint ID                              0            0.000000
Product                                   0            0.000000

From above we see that there are more than 50 percent missing values in five features.Generally it is good to drop the features that has more more than 50 percentage missing values as it does not contribute much to the model.But since we have five such features,instead of dropping them out we can create new features out of them by assigning 1 if there is a missing value and 0 if there is no missing value

In [91]:
#missing values in train plus test data...
missing_count=(combined_data).isnull().sum().sort_values(ascending=False)
missing_percentage=(combined_data).isnull().sum().sort_values(ascending=False)/len(combined_data)
pd.concat([missing_count,missing_percentage],axis=1,keys=['missing_count','missing_percentage'])

missing_count  missing_percentage
Tags                                 513950            0.859409
Consumer complaint narrative         504376            0.843400
Company public response              484859            0.810764
Consumer consent provided?           428676            0.716817
Sub-issue                            365685            0.611486
Sub-product                          173225            0.289661
Consumer disputed?                   119606            0.200001
ZIP code                               4774            0.007983
State                                  4764            0.007966
Submitted via                             1            0.000002
Issue                                     0            0.000000
day_difference                            0            0.000000
Company                                   0            0.000000
data                                      0            0.000000
Company response to consumer              0            0.000000
Timely response?                          0            0.000000
Complaint ID                              0            0.000000
Product                                   0            0.000000

since there are missing values even in the test it is not a bad idea to create new features based on the missing values

In [92]:
#creating new features based on the missing values
for col in ['Tags','Consumer complaint narrative','Company public response',
            'Consumer consent provided?','Sub-issue','Sub-product']:
    var_name=col.replace(' ','_').replace('?','').replace('-','_')+'_isNaN'
    combined_data[var_name]=np.where(combined_data[col].isnull(),1,0)
    combined_data.drop([col],axis=1,inplace=True)
    

In [93]:
combined_data.head(4)

Product                                     Issue  \
0              Credit card                         Billing statement   
1  Bank account or service  Making/receiving payments, sending money   
2         Credit reporting    Incorrect information on credit report   
3              Credit card                         Billing statement   

                 Company State ZIP code Submitted via  \
0  Wells Fargo & Company    MI    48342           Web   
1      Santander Bank US    PA    18042      Referral   
2                Equifax    CA    92427      Referral   
3           U.S. Bancorp    GA    305XX           Web   

      Company response to consumer Timely response? Consumer disputed?  \
0          Closed with explanation              Yes                 No   
1                           Closed              Yes                 No   
2  Closed with non-monetary relief              Yes                 No   
3      Closed with monetary relief              Yes                 No   

   Complaint ID   data    day_difference  Tags_isNaN  \
0        856103  train    86400000000000           0   
1       1034666  train   518400000000000           1   
2        756363  train  1814400000000000           1   
3       1474177  train                 0           0   

   Consumer_complaint_narrative_isNaN  Company_public_response_isNaN  \
0                                   1                              1   
1                                   1                              1   
2                                   1                              1   
3                                   0                              0   

   Consumer_consent_provided_isNaN  Sub_issue_isNaN  Sub_product_isNaN  
0                                1                1                  1  
1                                1                1                  0  
2                                1                0                  1  
3                                0                1                  1

In [94]:
combined_data['Product'].value_counts()

Mortgage                   195079
Debt collection            108047
Credit reporting           101612
Credit card                 71613
Bank account or service     68094
Consumer Loan               23263
Student loan                18635
Money transfers              4176
Payday loan                  4048
Prepaid card                 2813
Other financial service       635
Virtual currency               12
Name: Product, dtype: int64

In [95]:
combined_data['Product'].nunique()

12

In [96]:
combined_data.select_dtypes(['object']).columns

Index(['Product', 'Issue', 'Company', 'State', 'ZIP code', 'Submitted via',
       'Company response to consumer', 'Timely response?',
       'Consumer disputed?', 'data'],
      dtype='object')

In [97]:
for col in combined_data.select_dtypes(['object']).columns:
    print(col,':',combined_data[col].nunique())
    

Product : 12
Issue : 95
Company : 3437
State : 62
ZIP code : 27342
Submitted via : 6
Company response to consumer : 7
Timely response? : 2
Consumer disputed? : 2
data : 2


It's better to drop Company as it's not a good idea to make 3436 dummy variables.Also ZIP code does not contribute much to the target variable and hence it's goog to drop ZIP code

In [98]:
for col in ['Company','ZIP code']:
    combined_data.drop([col],axis=1,inplace=True)

In [99]:
combined_data.head(3)

Product                                     Issue State  \
0              Credit card                         Billing statement    MI   
1  Bank account or service  Making/receiving payments, sending money    PA   
2         Credit reporting    Incorrect information on credit report    CA   

  Submitted via     Company response to consumer Timely response?  \
0           Web          Closed with explanation              Yes   
1      Referral                           Closed              Yes   
2      Referral  Closed with non-monetary relief              Yes   

  Consumer disputed?  Complaint ID   data    day_difference  Tags_isNaN  \
0                 No        856103  train    86400000000000           0   
1                 No       1034666  train   518400000000000           1   
2                 No        756363  train  1814400000000000           1   

   Consumer_complaint_narrative_isNaN  Company_public_response_isNaN  \
0                                   1                              1   
1                                   1                              1   
2                                   1                              1   

   Consumer_consent_provided_isNaN  Sub_issue_isNaN  Sub_product_isNaN  
0                                1                1                  1  
1                                1                1                  0  
2                                1                0                  1

In [100]:
combined_data['Consumer disputed?']=np.where(combined_data['Consumer disputed?']=='Yes',1,0)

In [101]:
combined_data.head(3)

Product                                     Issue State  \
0              Credit card                         Billing statement    MI   
1  Bank account or service  Making/receiving payments, sending money    PA   
2         Credit reporting    Incorrect information on credit report    CA   

  Submitted via     Company response to consumer Timely response?  \
0           Web          Closed with explanation              Yes   
1      Referral                           Closed              Yes   
2      Referral  Closed with non-monetary relief              Yes   

   Consumer disputed?  Complaint ID   data    day_difference  Tags_isNaN  \
0                   0        856103  train    86400000000000           0   
1                   0       1034666  train   518400000000000           1   
2                   0        756363  train  1814400000000000           1   

   Consumer_complaint_narrative_isNaN  Company_public_response_isNaN  \
0                                   1                              1   
1                                   1                              1   
2                                   1                              1   

   Consumer_consent_provided_isNaN  Sub_issue_isNaN  Sub_product_isNaN  
0                                1                1                  1  
1                                1                1                  0  
2                                1                0                  1

In [102]:
combined_data.tail(4)

Product                                     Issue State  \
119602  Credit reporting  Credit reporting company's investigation    CA   
119603   Debt collection           Disclosure verification of debt    CA   
119604   Debt collection           Disclosure verification of debt    PA   
119605   Debt collection     Cont'd attempts collect debt not owed    CA   

       Submitted via Company response to consumer Timely response?  \
119602           Web      Closed with explanation              Yes   
119603         Phone      Closed with explanation              Yes   
119604         Phone      Closed with explanation              Yes   
119605           Web      Closed with explanation              Yes   

        Consumer disputed?  Complaint ID  data   day_difference  Tags_isNaN  \
119602                   0       2023523  test                0           1   
119603                   0       1413678  test  259200000000000           0   
119604                   0        732458  test   86400000000000           1   
119605                   0       1235226  test                0           1   

        Consumer_complaint_narrative_isNaN  Company_public_response_isNaN  \
119602                                   0                              0   
119603                                   1                              0   
119604                                   1                              1   
119605                                   1                              1   

        Consumer_consent_provided_isNaN  Sub_issue_isNaN  Sub_product_isNaN  
119602                                0                0                  1  
119603                                1                0                  0  
119604                                1                0                  0  
119605                                1                0                  0

In [103]:
combined_data['Timely response?']=np.where(combined_data['Timely response?']=='Yes',1,0)

In [104]:
combined_data.head(3)

Product                                     Issue State  \
0              Credit card                         Billing statement    MI   
1  Bank account or service  Making/receiving payments, sending money    PA   
2         Credit reporting    Incorrect information on credit report    CA   

  Submitted via     Company response to consumer  Timely response?  \
0           Web          Closed with explanation                 1   
1      Referral                           Closed                 1   
2      Referral  Closed with non-monetary relief                 1   

   Consumer disputed?  Complaint ID   data    day_difference  Tags_isNaN  \
0                   0        856103  train    86400000000000           0   
1                   0       1034666  train   518400000000000           1   
2                   0        756363  train  1814400000000000           1   

   Consumer_complaint_narrative_isNaN  Company_public_response_isNaN  \
0                                   1                              1   
1                                   1                              1   
2                                   1                              1   

   Consumer_consent_provided_isNaN  Sub_issue_isNaN  Sub_product_isNaN  
0                                1                1                  1  
1                                1                1                  0  
2                                1                0                  1

In [105]:
for col in ['Product','Submitted via','Company response to consumer']:
    temp=pd.get_dummies(combined_data[col],prefix=col,drop_first=True)
    combined_data=pd.concat([temp,combined_data],axis=1)

In [106]:
combined_data.head(4)

Company response to consumer_Closed with explanation  \
0                                                  1      
1                                                  0      
2                                                  0      
3                                                  0      

   Company response to consumer_Closed with monetary relief  \
0                                                  0          
1                                                  0          
2                                                  0          
3                                                  1          

   Company response to consumer_Closed with non-monetary relief  \
0                                                  0              
1                                                  0              
2                                                  1              
3                                                  0              

   Company response to consumer_Closed with relief  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   

   Company response to consumer_Closed without relief  \
0                                                  0    
1                                                  0    
2                                                  0    
3                                                  0    

   Company response to consumer_Untimely response  Submitted via_Fax  \
0                                               0                  0   
1                                               0                  0   
2                                               0                  0   
3                                               0                  0   

   Submitted via_Phone  Submitted via_Postal mail  Submitted via_Referral  \
0                    0                          0                       0   
1                    0                          0                       1   
2                    0                          0                       1   
3                    0                          0                       0   

   Submitted via_Web  Product_Consumer Loan  Product_Credit card  \
0                  1                      0                    1   
1                  0                      0                    0   
2                  0                      0                    0   
3                  1                      0                    1   

   Product_Credit reporting  Product_Debt collection  Product_Money transfers  \
0                         0                        0                        0   
1                         0                        0                        0   
2                         1                        0                        0   
3                         0                        0                        0   

   Product_Mortgage  Product_Other financial service  Product_Payday loan  \
0                 0                                0                    0   
1                 0                                0                    0   
2                 0                                0                    0   
3                 0                                0                    0   

   Product_Prepaid card  Product_Student loan  Product_Virtual currency  \
0                     0                     0                         0   
1                     0                     0                         0   
2                     0                     0                         0   
3                     0                     0                         0   

                   Product                                     Issue State  \
0              Credit card                         Billing statement    MI   
1  Bank account or service  Making/receiving payments, sending money    PA   
2       

In [107]:
combined_data.dtypes

Company response to consumer_Closed with explanation             uint8
Company response to consumer_Closed with monetary relief         uint8
Company response to consumer_Closed with non-monetary relief     uint8
Company response to consumer_Closed with relief                  uint8
Company response to consumer_Closed without relief               uint8
Company response to consumer_Untimely response                   uint8
Submitted via_Fax                                                uint8
Submitted via_Phone                                              uint8
Submitted via_Postal mail                                        uint8
Submitted via_Referral                                           uint8
Submitted via_Web                                                uint8
Product_Consumer Loan                                            uint8
Product_Credit card                                              uint8
Product_Credit reporting                                         uint8
Produc

In [108]:
combined_data['State'].value_counts()

CA    87499
FL    57765
TX    44853
NY    41219
GA    26668
NJ    24230
PA    21407
IL    21228
VA    19616
MD    19038
OH    18631
NC    16608
MI    15460
AZ    13303
WA    12475
MA    11912
CO    10229
TN     9188
MO     7732
SC     7547
NV     7345
OR     7140
MN     6976
CT     6914
IN     6597
WI     6540
AL     6122
LA     5885
KY     4292
OK     3960
DC     3524
UT     3212
DE     3110
NM     2993
NH     2971
KS     2903
MS     2744
AR     2574
IA     2499
ID     2079
HI     2075
ME     2048
RI     2001
NE     1857
WV     1562
PR     1430
VT     1002
MT      953
SD      799
AK      687
WY      586
ND      509
AE      244
AP      166
VI      154
GU       80
FM       35
MH       28
MP       20
AS       17
PW       11
AA       11
Name: State, dtype: int64

In [109]:
k=combined_data['State'].value_counts()

In [110]:
k.axes[0]

Index(['CA', 'FL', 'TX', 'NY', 'GA', 'NJ', 'PA', 'IL', 'VA', 'MD', 'OH', 'NC',
       'MI', 'AZ', 'WA', 'MA', 'CO', 'TN', 'MO', 'SC', 'NV', 'OR', 'MN', 'CT',
       'IN', 'WI', 'AL', 'LA', 'KY', 'OK', 'DC', 'UT', 'DE', 'NM', 'NH', 'KS',
       'MS', 'AR', 'IA', 'ID', 'HI', 'ME', 'RI', 'NE', 'WV', 'PR', 'VT', 'MT',
       'SD', 'AK', 'WY', 'ND', 'AE', 'AP', 'VI', 'GU', 'FM', 'MH', 'MP', 'AS',
       'PW', 'AA'],
      dtype='object')

In [111]:
for i in k.axes[0][0:15]:
    var_name='State_'+i.replace(',','_').replace(" ",'_')
    combined_data[var_name]=np.where(combined_data['State']==i,1,0)
    

In [112]:
combined_data.head(4)

Company response to consumer_Closed with explanation  \
0                                                  1      
1                                                  0      
2                                                  0      
3                                                  0      

   Company response to consumer_Closed with monetary relief  \
0                                                  0          
1                                                  0          
2                                                  0          
3                                                  1          

   Company response to consumer_Closed with non-monetary relief  \
0                                                  0              
1                                                  0              
2                                                  1              
3                                                  0              

   Company response to consumer_Closed with relief  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   

   Company response to consumer_Closed without relief  \
0                                                  0    
1                                                  0    
2                                                  0    
3                                                  0    

   Company response to consumer_Untimely response  Submitted via_Fax  \
0                                               0                  0   
1                                               0                  0   
2                                               0                  0   
3                                               0                  0   

   Submitted via_Phone  Submitted via_Postal mail  Submitted via_Referral  \
0                    0                          0                       0   
1                    0                          0                       1   
2                    0                          0                       1   
3                    0                          0                       0   

   Submitted via_Web  Product_Consumer Loan  Product_Credit card  \
0                  1                      0                    1   
1                  0                      0                    0   
2                  0                      0                    0   
3                  1                      0                    1   

   Product_Credit reporting  Product_Debt collection  Product_Money transfers  \
0                         0                        0                        0   
1                         0                        0                        0   
2                         1                        0                        0   
3                         0                        0                        0   

   Product_Mortgage  Product_Other financial service  Product_Payday loan  \
0                 0                                0                    0   
1                 0                                0                    0   
2                 0                                0                    0   
3                 0                                0                    0   

   Product_Prepaid card  Product_Student loan  Product_Virtual currency  \
0                     0                     0                         0   
1                     0                     0                         0   
2                     0                     0                         0   
3                     0                     0                         0   

                   Product                                     Issue State  \
0              Credit card                         Billing statement    MI   
1  Bank account or service  Making/receiving payments, sending money    PA   
2       

In [113]:
pd.set_option('display.max_columns',500)

In [114]:
combined_data.drop(['State'],axis=1,inplace=True)

In [115]:
combined_data.drop(['Company response to consumer','Product','Submitted via'],axis=1,inplace=True)

In [116]:
combined_data.head(4)

Company response to consumer_Closed with explanation  \
0                                                  1      
1                                                  0      
2                                                  0      
3                                                  0      

   Company response to consumer_Closed with monetary relief  \
0                                                  0          
1                                                  0          
2                                                  0          
3                                                  1          

   Company response to consumer_Closed with non-monetary relief  \
0                                                  0              
1                                                  0              
2                                                  1              
3                                                  0              

   Company response to consumer_Closed with relief  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   

   Company response to consumer_Closed without relief  \
0                                                  0    
1                                                  0    
2                                                  0    
3                                                  0    

   Company response to consumer_Untimely response  Submitted via_Fax  \
0                                               0                  0   
1                                               0                  0   
2                                               0                  0   
3                                               0                  0   

   Submitted via_Phone  Submitted via_Postal mail  Submitted via_Referral  \
0                    0                          0                       0   
1                    0                          0                       1   
2                    0                          0                       1   
3                    0                          0                       0   

   Submitted via_Web  Product_Consumer Loan  Product_Credit card  \
0                  1                      0                    1   
1                  0                      0                    0   
2                  0                      0                    0   
3                  1                      0                    1   

   Product_Credit reporting  Product_Debt collection  Product_Money transfers  \
0                         0                        0                        0   
1                         0                        0                        0   
2                         1                        0                        0   
3                         0                        0                        0   

   Product_Mortgage  Product_Other financial service  Product_Payday loan  \
0                 0                                0                    0   
1                 0                                0                    0   
2                 0                                0                    0   
3                 0                                0                    0   

   Product_Prepaid card  Product_Student loan  Product_Virtual currency  \
0                     0                     0                         0   
1                     0                     0                         0   
2                     0                     0                         0   
3                     0                     0                         0   

                                      Issue  Timely response?  \
0                         Billing statement                 1   
1  Making/receiving payments, sending money                 1   
2    Incorrect information on credit report    

In [117]:
combined_data['Issue'].value_counts()

Loan modification,collection,foreclosure    100244
Incorrect information on credit report       73345
Loan servicing, payments, escrow account     64286
Cont'd attempts collect debt not owed        45298
Account opening, closing, or management      29502
Disclosure verification of debt              20253
Communication tactics                        19114
Deposits and withdrawals                     18483
Application, originator, mortgage broker     14014
Billing disputes                             11962
Credit reporting company's investigation     11799
Other                                        11764
Managing the loan or lease                   11118
Problems caused by my funds being low         9651
Unable to get credit report/credit score      8926
False statements or representation            8836
Dealing with my lender or servicer            8064
Improper contact or sharing of info           7766
Problems when you are unable to pay           7362
Settlement process and costs   

In [118]:
pd.set_option('display.max_rows',500)

In [119]:
j=combined_data['Issue'].value_counts()
for i in j.axes[0][0:13]:
    var_name='Issue_'+i.replace(',','_').replace(' ','_')
    combined_data[var_name]=np.where(combined_data['Issue']==i,1,0)

In [120]:
combined_data.head(4)

Company response to consumer_Closed with explanation  \
0                                                  1      
1                                                  0      
2                                                  0      
3                                                  0      

   Company response to consumer_Closed with monetary relief  \
0                                                  0          
1                                                  0          
2                                                  0          
3                                                  1          

   Company response to consumer_Closed with non-monetary relief  \
0                                                  0              
1                                                  0              
2                                                  1              
3                                                  0              

   Company response to consumer_Closed with relief  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   

   Company response to consumer_Closed without relief  \
0                                                  0    
1                                                  0    
2                                                  0    
3                                                  0    

   Company response to consumer_Untimely response  Submitted via_Fax  \
0                                               0                  0   
1                                               0                  0   
2                                               0                  0   
3                                               0                  0   

   Submitted via_Phone  Submitted via_Postal mail  Submitted via_Referral  \
0                    0                          0                       0   
1                    0                          0                       1   
2                    0                          0                       1   
3                    0                          0                       0   

   Submitted via_Web  Product_Consumer Loan  Product_Credit card  \
0                  1                      0                    1   
1                  0                      0                    0   
2                  0                      0                    0   
3                  1                      0                    1   

   Product_Credit reporting  Product_Debt collection  Product_Money transfers  \
0                         0                        0                        0   
1                         0                        0                        0   
2                         1                        0                        0   
3                         0                        0                        0   

   Product_Mortgage  Product_Other financial service  Product_Payday loan  \
0                 0                                0                    0   
1                 0                                0                    0   
2                 0                                0                    0   
3                 0                                0                    0   

   Product_Prepaid card  Product_Student loan  Product_Virtual currency  \
0                     0                     0                         0   
1                     0                     0                         0   
2                     0                     0                         0   
3                     0                     0                         0   

                                      Issue  Timely response?  \
0                         Billing statement                 1   
1  Making/receiving payments, sending money                 1   
2    Incorrect information on credit report    

In [121]:
combined_data.drop(['Issue'],axis=1,inplace=True)

The data preparation part is done for both training data and test data.We now have to separate them and build the model using training data and predict the values for test data

In [122]:
from sklearn import preprocessing

In [123]:
combined_data['day_difference']=preprocessing.scale(combined_data['day_difference'])

C:\Users\chandrakanth.seela\Anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:174: UserWarning: Numerical issues were encountered when centering the data and might not be solved. Dataset may contain too large values. You may need to prescale your features.
  warnings.warn("Numerical issues were encountered "


In [124]:
combined_data.head(10)

Company response to consumer_Closed with explanation  \
0                                                  1      
1                                                  0      
2                                                  0      
3                                                  0      
4                                                  1      
5                                                  1      
6                                                  1      
7                                                  1      
8                                                  0      
9                                                  1      

   Company response to consumer_Closed with monetary relief  \
0                                                  0          
1                                                  0          
2                                                  0          
3                                                  1          
4                                                  0          
5                                                  0          
6                                                  0          
7                                                  0          
8                                                  1          
9                                                  0          

   Company response to consumer_Closed with non-monetary relief  \
0                                                  0              
1                                                  0              
2                                                  1              
3                                                  0              
4                                                  0              
5                                                  0              
6                                                  0              
7                                                  0              
8                                                  0              
9                                                  0              

   Company response to consumer_Closed with relief  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   
5                                                0   
6                                                0   
7                                                0   
8                                                0   
9                                                0   

   Company response to consumer_Closed without relief  \
0                                                  0    
1                                                  0    
2                                                  0    
3                                                  0    
4                                                  0    
5                                                  0    
6                                                  0    
7                                                  0    
8                                                  0    
9                                                  0    

   Company response to consumer_Untimely response  Submitted via_Fax  \
0                                               0                  0   
1                                               0                  0   
2                                               0                  0   
3                                               0                  0   
4                                               0                  0   
5                                               0                  0   
6                                               0                  0   
7                                               0                  0   
8                                               0 

In [125]:
consumer_train=combined_data.loc[combined_data['data']=='train']

In [126]:
consumer_train

Company response to consumer_Closed with explanation  \
0                                                       1      
1                                                       0      
2                                                       0      
3                                                       0      
4                                                       1      
...                                                   ...      
478416                                                  1      
478417                                                  1      
478418                                                  1      
478419                                                  1      
478420                                                  0      

        Company response to consumer_Closed with monetary relief  \
0                                                       0          
1                                                       0          
2                                                       0          
3                                                       1          
4                                                       0          
...                                                   ...          
478416                                                  0          
478417                                                  0          
478418                                                  0          
478419                                                  0          
478420                                                  0          

        Company response to consumer_Closed with non-monetary relief  \
0                                                       0              
1                                                       0              
2                                                       1              
3                                                       0              
4                                                       0              
...                                                   ...              
478416                                                  0              
478417                                                  0              
478418                                                  0              
478419                                                  0              
478420                                                  1              

        Company response to consumer_Closed with relief  \
0                                                     0   
1                                                     0   
2                                                     0   
3                                                     0   
4                                                     0   
...                                                 ...   
478416                                                0   
478417                                                0   
478418                                                0   
478419                                                0   
478420                                                0   

        Company response to consumer_Closed without relief  \
0                                                       0    
1                                                       0    
2                                                       0    
3                                                       0    
4                                                       0    
...                                                   ...    
478416                                                  0    
478417                                                  0    
478418                                                  0    
478419                                                  0    
478420                                                  0    

        Company response to consumer_Untimely response  Submitted via_Fax  \
0                         

In [127]:
consumer_test=combined_data.loc[combined_data['data']=='test']

In [164]:
consumer_test

Company response to consumer_Closed with explanation  \
0                                                       1      
1                                                       1      
2                                                       0      
3                                                       0      
4                                                       1      
...                                                   ...      
119601                                                  0      
119602                                                  1      
119603                                                  1      
119604                                                  1      
119605                                                  1      

        Company response to consumer_Closed with monetary relief  \
0                                                       0          
1                                                       0          
2                                                       0          
3                                                       0          
4                                                       0          
...                                                   ...          
119601                                                  1          
119602                                                  0          
119603                                                  0          
119604                                                  0          
119605                                                  0          

        Company response to consumer_Closed with non-monetary relief  \
0                                                       0              
1                                                       0              
2                                                       0              
3                                                       1              
4                                                       0              
...                                                   ...              
119601                                                  0              
119602                                                  0              
119603                                                  0              
119604                                                  0              
119605                                                  0              

        Company response to consumer_Closed with relief  \
0                                                     0   
1                                                     0   
2                                                     0   
3                                                     0   
4                                                     0   
...                                                 ...   
119601                                                0   
119602                                                0   
119603                                                0   
119604                                                0   
119605                                                0   

        Company response to consumer_Closed without relief  \
0                                                       0    
1                                                       0    
2                                                       1    
3                                                       0    
4                                                       0    
...                                                   ...    
119601                                                  0    
119602                                                  0    
119603                                                  0    
119604                                                  0    
119605                                                  0    

        Company response to consumer_Untimely response  Submitted via_Fax  \
0                         

In [161]:
del consumer_test['data']

In [163]:
del consumer_test['Consumer disputed?']

The data pre proessing part is done.Let's build a simple logistic regression with no penality(no regularization).To assess the performance of our model we split the training data into two parts,one to build the model and other to asess its performance.

In [129]:
from sklearn.model_selection import train_test_split

In [130]:
consumer_train_train,consumer_train_test = train_test_split(consumer_train,test_size=0.25,random_state=2)

In [131]:
consumer_train_train.shape

(358815, 61)

In [132]:
consumer_train_test.shape

(119606, 61)

In [133]:
x_train=consumer_train_train.drop(['Complaint ID','data','Consumer disputed?'],axis=1)
y_train=consumer_train_train['Consumer disputed?']
x_test=consumer_train_test.drop(['Complaint ID','data','Consumer disputed?'],axis=1)
y_test=consumer_train_test['Consumer disputed?']

In [134]:
x_train.head()

Company response to consumer_Closed with explanation  \
461347                                                  0      
284174                                                  1      
405314                                                  1      
335929                                                  1      
155301                                                  0      

        Company response to consumer_Closed with monetary relief  \
461347                                                  0          
284174                                                  0          
405314                                                  0          
335929                                                  0          
155301                                                  1          

        Company response to consumer_Closed with non-monetary relief  \
461347                                                  1              
284174                                                  0              
405314                                                  0              
335929                                                  0              
155301                                                  0              

        Company response to consumer_Closed with relief  \
461347                                                0   
284174                                                0   
405314                                                0   
335929                                                0   
155301                                                0   

        Company response to consumer_Closed without relief  \
461347                                                  0    
284174                                                  0    
405314                                                  0    
335929                                                  0    
155301                                                  0    

        Company response to consumer_Untimely response  Submitted via_Fax  \
461347                                               0                  0   
284174                                               0                  0   
405314                                               0                  0   
335929                                               0                  0   
155301                                               0                  0   

        Submitted via_Phone  Submitted via_Postal mail  \
461347                    0                          0   
284174                    0                          0   
405314                    0                          0   
335929                    0                          0   
155301                    0                          0   

        Submitted via_Referral  Submitted via_Web  Product_Consumer Loan  \
461347                       0                  1                      0   
284174                       0                  1                      0   
405314                       1                  0                      0   
335929                       1                  0                      0   
155301                       1                  0                      0   

        Product_Credit card  Product_Credit reporting  \
461347                    0                         1   
284174                    0                         1   
405314                    0                         0   
335929                    0                         0   
155301                    1                         0   

        Product_Debt collection  Product_Money transfers  Product_Mortgage  \
461347                        0                        0                 0   
284174                        0                        0                 0   
405314                        0                        0                 1   
335929                        0                        0                 1   
155301                        0                    

In [135]:
y_train.value_counts()

0    282570
1     76245
Name: Consumer disputed?, dtype: int64

In [136]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

In [137]:
logr=LogisticRegression(penalty='none',class_weight="balanced",max_iter=1000,random_state=2)

In [138]:
logr.fit(x_train,y_train)

LogisticRegression(class_weight='balanced', max_iter=1000, penalty='none',
                   random_state=2)

In [139]:
# score model performance on the test data
roc_auc_score(y_test,logr.predict(x_test))

0.5806025201730223

The basic model is built and we see the roc_auc_score is 0.5806.Now let's apply regularization and do hyperparameter tunig combined with cross validation to obtain the best possible accurate model.

For cross validation we use the entire training data.


In [140]:
X_Train=consumer_train.drop(['Complaint ID','data','Consumer disputed?'],axis=1)
Y_Train=consumer_train['Consumer disputed?']

In [192]:
Y_Train.head(10)

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    1
8    0
9    0
Name: Consumer disputed?, dtype: int32

In [141]:
params={'class_weight':['balanced',None],
        'penalty':['l1','l2'],
        'C':np.linspace(0.01,1000,10)}

In [142]:
model=LogisticRegression(fit_intercept=True,solver='liblinear',max_iter=1000)

In [143]:
from sklearn.model_selection import GridSearchCV
from sklearn.exceptions import FitFailedWarning

In [144]:
grid_search=GridSearchCV(model,param_grid=params,cv=3,scoring="roc_auc")

In [145]:
grid_search.fit(X_Train,Y_Train)

GridSearchCV(cv=3,
             estimator=LogisticRegression(max_iter=1000, solver='liblinear'),
             param_grid={'C': array([1.0000e-02, 1.1112e+02, 2.2223e+02, 3.3334e+02, 4.4445e+02,
       5.5556e+02, 6.6667e+02, 7.7778e+02, 8.8889e+02, 1.0000e+03]),
                         'class_weight': ['balanced', None],
                         'penalty': ['l1', 'l2']},
             scoring='roc_auc')

In [222]:
def report(results, n_top):
    for i in range(1, n_top + 1):
        candidates = np.flatnonzero(results['rank_test_score'] == i)
        for candidate in candidates:
            print("Model with rank: {0}".format(i))
            print("Mean validation score: {0:.3f} (std: {1:.3f})".format(
                  results['mean_test_score'][candidate],
                  results['std_test_score'][candidate]))
            print("Parameters: {0}".format(results['params'][candidate]))
            print("")

In [223]:
report(grid_search.cv_results_,5)

Model with rank: 1
Mean validation score: 0.613 (std: 0.001)
Parameters: {'C': 1000.0, 'class_weight': None, 'penalty': 'l1'}

Model with rank: 2
Mean validation score: 0.613 (std: 0.001)
Parameters: {'C': 555.56, 'class_weight': None, 'penalty': 'l2'}

Model with rank: 3
Mean validation score: 0.613 (std: 0.001)
Parameters: {'C': 1000.0, 'class_weight': None, 'penalty': 'l2'}

Model with rank: 4
Mean validation score: 0.613 (std: 0.001)
Parameters: {'C': 111.12, 'class_weight': None, 'penalty': 'l2'}

Model with rank: 5
Mean validation score: 0.613 (std: 0.001)
Parameters: {'C': 666.67, 'class_weight': None, 'penalty': 'l2'}



Here due to run time constraint cross-validation with k value as 3 is implemented.That gives a total of  2 x 2 x 10 x 3 = 120 models.( 2 penalty values, 2 class_weights, 10 C values give 40 combinations and each combination is implemented with cross-validation validation value of 3 i.e 40 x 3 = 120 combinations)
Of all these models the best possible roc_auc_score obtained is 0.613 with Parameters: {'C': 1000.0, 'class_weight': None, 'penalty': 'l1'}.If we increase the cross-validation value to let's say 10 the model's performance may increase but it comes at the expense of implementng 2 x 2 x 10 x 10 = 400 possible combinations of models  which may take a huge run time.

In [156]:
grid_search.best_estimator_

LogisticRegression(C=1000.0, max_iter=1000, penalty='l1', solver='liblinear')

In [157]:
logreg=grid_search.best_estimator_

In [154]:
logreg.fit(X_Train,Y_Train)

LogisticRegression(C=1000.0, max_iter=1000, penalty='l1', solver='liblinear')

If we need to predict the probabilties

In [178]:
test_score=logreg.predict_proba(consumer_test.drop(['Complaint ID'],axis=1))


In [179]:
test_score

array([[0.70829149, 0.29170851],
       [0.80116036, 0.19883964],
       [0.66471652, 0.33528348],
       ...,
       [0.78133751, 0.21866249],
       [0.79761189, 0.20238811],
       [0.76899478, 0.23100522]])

In [180]:
logreg.classes_

array([0, 1])

In [181]:
test_score=logreg.predict_proba(consumer_test.drop(['Complaint ID'],axis=1))[:,1]

In [182]:
test_score

array([0.29170851, 0.19883964, 0.33528348, ..., 0.21866249, 0.20238811,
       0.23100522])

In [183]:
pd.DataFrame(test_score).to_csv("consumer_probability_submission.csv",index=False)

Let's look at the co-efficients of the features

In [187]:
logreg.intercept_

array([-0.5482248])

In [188]:
list(zip(x_train.columns,logr.coef_[0]))

[('Company response to consumer_Closed with explanation', 0.03799603537211484),
 ('Company response to consumer_Closed with monetary relief',
  -0.9261519318282542),
 ('Company response to consumer_Closed with non-monetary relief',
  -0.6226651831817984),
 ('Company response to consumer_Closed with relief', -0.6870889819996964),
 ('Company response to consumer_Closed without relief', 0.18987598692505145),
 ('Company response to consumer_Untimely response', 0.5320847159107176),
 ('Submitted via_Fax', 0.049434635404542855),
 ('Submitted via_Phone', -0.26248141806350217),
 ('Submitted via_Postal mail', -0.3234986764841859),
 ('Submitted via_Referral', -0.2736315827073786),
 ('Submitted via_Web', 0.17270753830411562),
 ('Product_Consumer Loan', 0.03402748242845818),
 ('Product_Credit card', 0.06525956895855697),
 ('Product_Credit reporting', -0.032145055620478036),
 ('Product_Debt collection', 0.011813311308899043),
 ('Product_Money transfers', -0.4842653592371014),
 ('Product_Mortgage', 0

If we need to submit the classes we need to choose a threshold value based on which we decide the class of the predicted value 

In [189]:
cutoffs=np.linspace(0.01,0.99,99)

cutoffs

array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 , 0.11,
       0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 , 0.21, 0.22,
       0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 , 0.31, 0.32, 0.33,
       0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 , 0.41, 0.42, 0.43, 0.44,
       0.45, 0.46, 0.47, 0.48, 0.49, 0.5 , 0.51, 0.52, 0.53, 0.54, 0.55,
       0.56, 0.57, 0.58, 0.59, 0.6 , 0.61, 0.62, 0.63, 0.64, 0.65, 0.66,
       0.67, 0.68, 0.69, 0.7 , 0.71, 0.72, 0.73, 0.74, 0.75, 0.76, 0.77,
       0.78, 0.79, 0.8 , 0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87, 0.88,
       0.89, 0.9 , 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99])

In [190]:
train_score=logr.predict_proba(X_Train)[:,1]

real=Y_Train

In [203]:
train_score

array([0.60128126, 0.44191962, 0.23702727, ..., 0.47109016, 0.60549725,
       0.42481903])

Let's select the threshold value of probability.Here the problem we are trying to solve is to mininmize the False Negatives,i.e to not miss the consumers who are likely to dispute.So the relavant evaluation metric would be recall.
we will use F-Beta score with higher value of beta to focus more on recall

In [261]:
F_beta=[]

for cutoff in cutoffs:
    
    predicted=(train_score>cutoff).astype(int)

    TP=((predicted==1) & (real==1)).sum()
    TN=((predicted==0) & (real==0)).sum()
    FP=((predicted==1) & (real==0)).sum()
    FN=((predicted==0) & (real==1)).sum()
    
    Precision = TP / (TP+FP)
    Recall = TP / (TP+FN)
      
    f_beta_score = (5*Precision*Recall) / ((4*Precision) + Recall) #beta=2
    
    
    F_beta.append(f_beta_score)



<ipython-input-261-1bcf3ad3eefe>:12: RuntimeWarning: invalid value encountered in longlong_scalars
  Precision = TP / (TP+FP)


In [262]:
list(zip(cutoffs,F_beta))

[(0.01, 0.5736106634092825),
 (0.02, 0.5736106634092825),
 (0.03, 0.5736113121844405),
 (0.04, 0.5736113121844405),
 (0.05, 0.5736113121844405),
 (0.060000000000000005, 0.5736113121844405),
 (0.06999999999999999, 0.5736126097391596),
 (0.08, 0.57361585365164),
 (0.09, 0.57361585365164),
 (0.09999999999999999, 0.5736140911245053),
 (0.11, 0.573612977365166),
 (0.12, 0.5736112147925955),
 (0.13, 0.5736164051984483),
 (0.14, 0.5736178866846361),
 (0.15000000000000002, 0.5736177894811854),
 (0.16, 0.5736203744762493),
 (0.17, 0.5736568026183455),
 (0.18000000000000002, 0.5736982463762705),
 (0.19, 0.573784869575058),
 (0.2, 0.5739546471137786),
 (0.21000000000000002, 0.5742298823993525),
 (0.22, 0.5747234798490696),
 (0.23, 0.5750680005007567),
 (0.24000000000000002, 0.575388068331503),
 (0.25, 0.5760354691075515),
 (0.26, 0.5766923135277039),
 (0.27, 0.5770014159614747),
 (0.28, 0.5771743739095982),
 (0.29000000000000004, 0.5775475526897517),
 (0.3, 0.5781566203577738),
 (0.31, 0.57842419

In [263]:
mycutoff=cutoffs[F_beta==max(F_beta)][0]
mycutoff

0.31

In [264]:
test_classes=(test_score>mycutoff).astype(int)

In [265]:
test_classes

array([0, 0, 1, ..., 0, 0, 0])

In [266]:
pd.DataFrame(test_classes).to_csv("consumer_classes_submission.csv",index=False)

In [267]:
(test_score>mycutoff).sum()

7041

In [268]:
test_score

array([0.29170851, 0.19883964, 0.33528348, ..., 0.21866249, 0.20238811,
       0.23100522])

In [269]:
prediction=np.where((test_score>mycutoff).astype(int) == 1,"Yes","No")
submission=pd.DataFrame(list(zip(consumer_test['Complaint ID'],list(prediction))),
                       columns=['Complaint ID','Consumer disputed?'])

In [270]:
submission.head(10)

Complaint ID Consumer disputed?
0        675956                 No
1       1858795                 No
2         32637                Yes
3       1731374                 No
4        501487                 No
5        631565                 No
6       1283963                 No
7       1902970                 No
8       1160691                 No
9        359929                 No

In [271]:
submission['Consumer disputed?'].value_counts()

No     112565
Yes      7041
Name: Consumer disputed?, dtype: int64